forecasting/src/autogluon/forecasting/trainer/abstract_trainer.py

```diff
@@ -0,0 +1,668 @@
+
+    def _train_single(
+        self,
+        train_data: TimeSeriesDataFrame,
+        model: AbstractForecastingModel,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        time_limit: Optional[float] = None,
+    ) -> AbstractForecastingModel:
+        """Train the single model and return the model object that was fitted. This method
+        does not save the resulting model."""
+        model.fit(train_data=train_data, val_data=val_data, time_limit=time_limit)
+        return model
+
+    def _train_single_full(
+        self,
+        train_data: TimeSeriesDataFrame,
+        model: AbstractForecastingModel,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        hyperparameter_tune: bool = False,
+        time_limit: Optional[float] = None,
+    ) -> List[str]:
+        if hyperparameter_tune:
+            scheduler_cls, scheduler_options = scheduler_factory("auto")
+            scheduler_options["num_trials"] = 10 if time_limit is None else None
+            hpo_models, hpo_model_performances, hpo_results = model.hyperparameter_tune(
+                train_data=train_data,
+                val_data=val_data,
+                scheduler_options=(scheduler_cls, scheduler_options),
+                time_limit=time_limit,
+            )
+            hpo_results.pop("search_strategy", None)
+            self.hpo_results[model.name] = hpo_results
+            model_names_trained = []
+
+            # add each of the trained HPO configurations to the trained models
+            for model_hpo_name, model_path in hpo_models.items():
+                model_hpo = self.load_model(
+                    model_hpo_name, path=model_path, model_type=type(model)
+                )
+                self._add_model(model_hpo)
+                model_names_trained.append(model_hpo.name)
+        else:
+            model_names_trained = self._train_and_save(
+                train_data, model=model, val_data=val_data, time_limit=time_limit
+            )
+
+        return model_names_trained
+
+    def _train_and_save(
+        self,
+        train_data: TimeSeriesDataFrame,
+        model: AbstractForecastingModel,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        time_limit: Optional[float] = None,
+    ) -> List[str]:
+        """Fit and save the given model on given training and validation data and save the
+        trained model.
+
+        Returns
+        -------
+        model_names_trained: the list of model names that were successfully trained
+        """
+        fit_start_time = time.time()
+        model_names_trained = []
+        try:
+            if time_limit is not None:
+                if time_limit <= 0:
+                    logging.log(
+                        15, f"Skipping {model.name} due to lack of time remaining."
+                    )
+                    return model_names_trained
+            else:
+                logging.log(20, f"Fitting model: {model.name} ...")
```

Comment:
logic like this is pretty generic, I wonder if we can have a clever way of reusing this between the different modules.

Reply:
we could probably employ a decorator like the following. I'll propose to push this to a different PR?

```python
import functools
import logging

class TimeLimitExceeded(Exception):
    pass

def time_limited(*decorator_args, **decorator_kwargs):
    
    def decorator(func):
    
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            time_limit = kwargs.pop("time_limit", None)
            
            if time_limit is not None:
                if time_limit <= 0:
                    logging.log(
                        15, f"Skipping due to lack of time remaining."
                    )
                    raise TimeLimitExceeded()
                else:
                    logging.log(20, f"Fitting model: ...")

            
            return func(*args, **kwargs)
        
        return wrapper
    
    return decorator

@time_limited()
def my_wrapped_function(a, b):
    print(a + b)
    
my_wrapped_function(3, 4, time_limit=-5)
```